src/services/BadgesService.ts

```diff
@@ -52,4 +64,74 @@ export class BadgesService {
   private static getIpfsHttpsLink(ipfsLink: string) {
     return ipfsLink.replace('ipfs://', 'https://ipfs.io/ipfs/')
   }
+
+  public static async giveBadgeToUsers(badgeCid: string, users: string[]): Promise<AirdropOutcome> {
+    try {
+      const badgeOwners = await OtterspaceSubgraph.get().getBadgeOwners(badgeCid)
+      const usersWithoutBadge = users.filter((user) => {
+        return !badgeOwners.includes(user.toLowerCase())
+      })
+
+      if (usersWithoutBadge.length > 0) {
+        return await this.airdropWithRetry(badgeCid, usersWithoutBadge)
+      } else {
+        return { status: AirdropJobStatus.FAILED, error: 'All recipients already have this badge' }
+      }
+    } catch (e) {
+      return { status: AirdropJobStatus.FAILED, error: JSON.stringify(e) }
+    }
+  }
+
+  private static async airdropWithRetry(
+    badgeCid: string,
+    recipients: string[],
+    retries = 3,
+    pumpGas = false
+  ): Promise<AirdropOutcome> {
+    try {
+      await airdrop(badgeCid, recipients, pumpGas)
+      return { status: AirdropJobStatus.FINISHED, error: '' }
+    } catch (error: any) {
+      if (retries > 0) {
+        logger.log(`Retrying airdrop... Attempts left: ${retries}`, error)
+        const pumpGas = this.isTransactionUnderpricedError(error)
+        return await this.airdropWithRetry(badgeCid, recipients, retries - 1, pumpGas)
+      } else {
+        logger.error('Airdrop failed after maximum retries', error)
+        return { status: AirdropJobStatus.FAILED, error: JSON.stringify(error) }
+      }
+    }
+  }
+
+  private static isTransactionUnderpricedError(error: any) {
+    try {
+      const errorParsed = JSON.parse(error.body)
+      const errorCode = errorParsed?.error?.code
+      return errorCode === TRANSACTION_UNDERPRICED
+    } catch (e) {
+      return false
+    }
+  }
+
+  static async giveLegislatorBadges(acceptedProposals: ProposalAttributes[]) {
+    const governanceProposals = acceptedProposals.filter((proposal) => proposal.type === ProposalType.Governance)
+    const coauthors = await CoauthorModel.findAllCoauthors(governanceProposals, CoauthorStatus.APPROVED)
+    const authors = governanceProposals.map((proposal) => proposal.user)
+    const authorsAndCoauthors = new Set([...authors.map(getChecksumAddress), ...coauthors.map(getChecksumAddress)])
+    await this.queueAirdropJob(LEGISLATOR_BADGE_SPEC_CID, Array.from(authorsAndCoauthors))
```

Comment:
If this is going to a queue, do we need to await it?

Reply:
Yes, if we don't await and something goes wrong while saving the job to the DB, we won't know (it happened to me while testing)